activerecord/lib/active_record/persistence.rb

```diff
@@ -709,6 +709,12 @@ def increment!(attribute, by = 1, touch: nil)
       increment(attribute, by)
       change = public_send(attribute) - (attribute_in_database(attribute.to_s) || 0)
       self.class.update_counters(id, attribute => change, touch: touch)
+
+      if locking_enabled?
+        self[self.class.locking_column] += 1
+        clear_attribute_change(self.class.locking_column)
+      end
```

Comment:
Since `Persistence` shouldn't necessarily be concerned with locking, what about moving this to a `Locking::Optimistic#increment!` override?  Something similar to [`_create_record`](https://github.com/rails/rails/blob/c360644ed90a350d75b1722837ec306b38268ab5/activerecord/lib/active_record/locking/optimistic.rb#L64-L71).

Reply:
@jonathanhefner - that's even better, I've rebased accordingly.

(Congrats on joining the Rails team, btw!)